### PR TITLE
move pybind11/Python api to cpython

### PR DESCRIPTION
**Pull request by marigoold**
此PR完成了：
- 把部分Tensor api从Python/Pybind11转移到了CPython中
- 增加了tp_hash函数，即 `Tensor.__hash__()` 。

关于hash函数，原来oneflow中没有手动指定，因此继承了 `object.__hash__()` ，如下所示：
```python
In [1]: import oneflow as flow

In [2]: a = flow.randn(3, 3)

In [3]: id(object.__hash__)
Out[3]: 139710681759968

In [4]: id(flow.Tensor.__hash__)
Out[4]: 139710681759968

In [5]: flow.Tensor.__hash__ == object.__hash__
Out[5]: True

In [6]: a.__hash__()
Out[6]: 8731912099991

In [7]: object.__hash__(a)
Out[7]: 8731912099991
```

由于Python中只指定 `__eq__` 而不指定 `__hash__` 时，会将 `__hash__` 设置成None，所以PR里手动把Tensor的 `__hash__` 设置成了object相同的函数。设置后效果如下：

```python
In [1]: import oneflow as flow; a = flow.randn(3, 3)

In [2]: a.__hash__()
Out[2]: 8745811753919

In [3]: object.__hash__(a)
Out[3]: 8745811753919
```

在迁移过程中发现部分缺失的api但torch已经实现：
rmod、rfloordiv、positive、cos_

发现部分api有bug：
negative 会强制把结果的dtype转成float32，而不是和输入tensor的dtype相同

这个部分迁移的api如下：
- [x]  Tensor.__mul__ = **lambda** self, other: self.mul(other)
- [x]  Tensor.__rmul__ = **lambda** self, other: self.mul(other)
- [x]  Tensor.__add__ = lambda self, other: self.add(other)
- [ ]  Tensor.__iadd__ = **lambda** self, other: self.add_(other)
- [x]  Tensor.__matmul__ = **lambda** self, other: self.matmul(other)
- [x]  Tensor.byte = _byte
- [ ]  Tensor.ndim = property(_ndim)
- [ ]  Tensor.numpy = _numpy
- [ ]  Tensor.size = _size
- [x]  Tensor.dim = _ndim
- [x]  Tensor.ndimension = _ndim
- [x]  Tensor.nelement = _nelement
- [x]  Tensor.numel = _numel
- [x]  Tensor.element_size = _element_size
- [ ]  Tensor.backward = _backward
- [ ]  Tensor.__setitem__ = _setitem
- [ ]  Tensor.__str__ = _str
- [ ]  Tensor.__repr__ = _repr
- [x]  Tensor.__eq__ = _eq
- [x]  Tensor.__ne__ = _ne
- [ ]  Tensor.__bool__ = is_nonzero
- [x]  Tensor.__gt__ = _gt
- [x]  Tensor.__lt__ = _lt
- [x]  Tensor.__ge__ = _ge
- [x]  Tensor.__le__ = _le
- [x]  Tensor.__and__ = _and
- [x]  Tensor.__or__ = _or
- [x]  Tensor.__xor__ = _xor
- [x]  Tensor.__mul__ = _mul
- [x]  Tensor.__rmul__ = _rmul
- [x]  *# Tensor.__add__ = _add*
- [ ]  Tensor.__iadd__ = _iadd
- [x]  *# Tensor.__radd__ = _radd*
- [ ]  Tensor.addmm = _addmm
- [x]  Tensor.__sub__ = _sub
- [x]  Tensor.__rsub__ = _rsub
- [x]  Tensor.__truediv__ = _truediv
- [x]  Tensor.__rtruediv__ = _rtruediv
- [x]  Tensor.__neg__ = _neg
- [x]  Tensor.__pow__ = _pow
- [x]  Tensor.__rpow__ = _rpow
- [ ]  Tensor.__format__ = _format
- [x]  Tensor.__floordiv__ = _floor_divide
- [x]  Tensor.__mod__ = _fmod
- [ ]  Tensor.__index__ = _index
- [x]  Tensor.__invert__ = _invert
- [ ]  Tensor.__float__ = _scalar_float
- [ ]  Tensor.__int__ = _scalar_int
- [ ]  Tensor.__array__ = _numpy
- [ ]  Tensor.uniform_ = _uniform
- [ ]  Tensor.trunc_normal_ = _trunc_normal_
- [ ]  Tensor.kaiming_uniform_ = _kaiming_uniform
- [ ]  Tensor.kaiming_normal_ = _kaiming_normal
- [ ]  Tensor.xavier_normal_ = _xavier_normal
- [ ]  Tensor.xavier_uniform_ = _xavier_uniform
- [ ]  Tensor.orthogonal_ = _orthogonal
- [ ]  Tensor.normal_ = _normal
- [ ]  Tensor.fill_ = _fill
- [ ]  Tensor.copy_ = _copy
- [x]  Tensor.get_device = _get_device
- [ ]  Tensor._meta_repr = _meta_repr
- [x]  Tensor.abs = _abs
- [x]  Tensor.exp = _exp
- [ ]  Tensor.floor_divide = _floor_divide
- [x]  Tensor.floor = _floor
- [x]  Tensor.floor_ = _floor_inplace_
- [ ]  Tensor.argmax = _argmax
- [ ]  Tensor.argmin = _argmin
- [ ]  Tensor.argsort = _argsort
- [ ]  Tensor.argwhere = _argwhere
- [x]  Tensor.acos = _acos
- [x]  Tensor.arccos = _arccos
- [x]  Tensor.acosh = _acosh
- [ ]  Tensor.amin = _amin
- [x]  Tensor.arccosh = _arccosh
- [x]  Tensor.atanh = _atanh
- [ ]  Tensor.atan2 = _atan2
- [x]  Tensor.arctanh = _arctanh
- [x]  Tensor.sign = _sign
- [x]  Tensor.sinh = _sinh
- [x]  Tensor.tan = _tan
- [ ]  [Tensor.gt](http://tensor.gt/) = _gt
- [ ]  [Tensor.ge](http://tensor.ge/) = _ge
- [x]  Tensor.gelu = _gelu
- [x]  Tensor.mish = _mish
- [x]  Tensor.negative = _negative
- [x]  Tensor.neg = _neg
- [x]  Tensor.sigmoid = _sigmoid
- [x]  Tensor.tanh = _tanh
- [x]  Tensor.silu = _silu
- [x]  Tensor.selu = _selu
- [x]  Tensor.softsign = _softsign
- [ ]  Tensor.cast = _cast
- [ ]  Tensor.diag = _diag
- [ ]  Tensor.diagonal = _diagonal
- [x]  Tensor.log1p = _log1p
- [x]  Tensor.log2 = _log2
- [ ]  Tensor.add = _add
- [ ]  Tensor.add_ = _add_inplace
- [ ]  Tensor.addcmul = _addcmul
- [ ]  Tensor.addcmul_ = _addcmul_
- [ ]  Tensor.div = _truediv
- [ ]  Tensor.div_ = _truediv_inplace
- [ ]  Tensor.mul = _mul
- [ ]  Tensor.mul_ = _mul_
- [x]  Tensor.reciprocal = _reciprocal
- [ ]  Tensor.sub = _sub
- [ ]  Tensor.sub_ = _sub_inplace
- [x]  Tensor.asin = _asin
- [x]  Tensor.arcsin = _arcsin
- [x]  Tensor.asinh = _asinh
- [x]  Tensor.arcsinh = _arcsinh
- [x]  Tensor.atan = _atan
- [x]  Tensor.arctan = _arctan
- [x]  Tensor.ceil = _ceil
- [ ]  Tensor.clamp = _clamp
- [ ]  Tensor.clamp_ = _clamp_
- [ ]  Tensor.clip = _clip
- [ ]  Tensor.clip_ = _clip_
- [x]  Tensor.cos = _cos
- [x]  Tensor.cosh = _cosh
- [ ]  Tensor.cpu = _cpu
- [ ]  Tensor.cuda = _cuda
- [ ]  Tensor.expand = _expand
- [ ]  Tensor.expand_as = _expand_as
- [x]  Tensor.erf = _erf
- [x]  Tensor.erfc = _erfc
- [x]  Tensor.erfinv = _erfinv
- [x]  Tensor.erfinv_ = _erfinv_inplace
- [x]  Tensor.expm1 = _expm1
- [ ]  Tensor.fmod = _fmod
- [ ]  Tensor.flatten = _flatten
- [ ]  Tensor.flip = _flip
- [ ]  Tensor.in_top_k = _in_top_k
- [ ]  Tensor.index_select = _index_select
- [x]  Tensor.log = _log
- [ ]  Tensor.minimum = _minimum
- [ ]  Tensor.maximum = _maximum
- [ ]  Tensor.new_empty = _new_empty
- [ ]  Tensor.new_ones = _new_ones
- [ ]  Tensor.new_zeros = _new_zeros
- [ ]  Tensor.pow = _pow
- [x]  Tensor.rsqrt = _rsqrt
- [x]  Tensor.sqrt = _sqrt
- [x]  Tensor.square = _square
- [ ]  Tensor.var = _var
- [ ]  Tensor.std = _std
- [ ]  Tensor.matmul = _matmul
- [x]  Tensor.round = _round
- [ ]  Tensor.softplus = _softplus
- [ ]  Tensor.tril = _tril
- [ ]  Tensor.triu = _triu
- [ ]  Tensor.where = _where
- [ ]  Tensor.norm = _norm
- [ ]  Tensor.transpose = _transpose
- [ ]  Tensor.permute = _permute
- [ ]  Tensor.local_to_global = _local_to_global
- [ ]  Tensor.global_to_global = _global_to_global
- [ ]  Tensor.to_global = _to_global
- [ ]  Tensor.relu = _relu
- [ ]  Tensor.relu_ = _relu_inplace
- [ ]  Tensor.softmax = _softmax
- [ ]  Tensor.log_softmax = _log_softmax
- [ ]  Tensor.logical_and = _and
- [ ]  Tensor.logical_or = _or
- [ ]  Tensor.logical_not = _not
- [ ]  Tensor.logical_xor = _xor
- [ ]  Tensor.roll = _roll
- [ ]  Tensor.bmm = _bmm
- [ ]  Tensor.chunk = _chunk
- [ ]  Tensor.repeat = _repeat
- [ ]  Tensor.tile = _tile
- [ ]  Tensor.split = _split
- [ ]  Tensor.unbind = _unbind
- [ ]  Tensor.squeeze = _squeeze
- [ ]  Tensor.swapaxes = _swapaxes
- [ ]  Tensor.amax = _amax
- [ ]  Tensor.swapdims = _swapdims
- [ ]  Tensor.unfold = _unfold
- [ ]  Tensor.narrow = _narrow
- [ ]  Tensor.unsqueeze = _unsqueeze
- [ ]  [Tensor.to](http://tensor.to/) = _to
- [ ]  Tensor.half = _half
- [ ]  Tensor.gather = _gather
- [ ]  Tensor.all = _all
- [ ]  Tensor.any = _any
- [ ]  Tensor.T = property(_T)
- [x]  Tensor.t = _t
- [ ]  Tensor.masked_fill = _masked_fill
- [ ]  Tensor.masked_select = _masked_select
- [ ]  Tensor.eq = _eq
- [ ]  [Tensor.ne](http://tensor.ne/) = _ne
- [ ]  Tensor.item = _item
- [ ]  [Tensor.lt](http://tensor.lt/) = _lt
- [ ]  Tensor.le = _le
- [ ]  Tensor.to_local = _to_local
- [ ]  Tensor.reshape = _reshape
- [ ]  Tensor.reshape_as = _reshape_as
- [ ]  Tensor.view = _view
- [ ]  Tensor.view_as = _view_as
- [ ]  Tensor.sort = _sort
- [ ]  Tensor.type_as = _type_as
- [ ]  Tensor.tolist = _tolist
- [ ]  Tensor.int = _int
- [ ]  Tensor.long = _long
- [ ]  Tensor.float = _float
- [ ]  Tensor.double = _double
- [ ]  Tensor.is_floating_point = _is_floating_point
- [ ]  Tensor.topk = _topk
- [ ]  Tensor.nms = _nms
- [ ]  Tensor.nonzero = _nonzero
- [ ]  Tensor.max = _max
- [ ]  Tensor.min = _min
- [ ]  Tensor.median = _median
- [ ]  Tensor.sum = _sum
- [ ]  Tensor.mean = _mean
- [ ]  Tensor.prod = _prod
- [x]  Tensor.sin = _sin
- [x]  Tensor.sin_ = _sin_inplace
- [ ]  Tensor.is_consistent = _is_consistent
- [ ]  Tensor.to_consistent = _to_consistent
- [x]  Tensor.isnan = _isnan
- [x]  Tensor.isinf = _isinf
- [ ]  Tensor.new_tensor = _new_tensor
- [ ]  Tensor.cumsum = _cumsum
- [ ]  Tensor.cumprod = _cumprod

### CODE REVIEW

File: oneflow/api/python/framework/tensor_functions.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#define ONEFLOW_API_PYTHON_FRAMEWORK_TENSOR_FUNCTIONS_H
+
+#include <Python.h>
+
+namespace oneflow {
+namespace one {
+
+extern PyNumberMethods PyTensorObject_as_number;
+extern PyMethodDef PyTensorObject_extra_methods[];
+extern PyObject* PyTensorObject_richcompare(PyObject*, PyObject*, int);
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.h:26]: 这里不用加extern

File: oneflow/api/python/framework/tensor_functions.cpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+namespace oneflow {
+namespace one {
+
+#define ASSERT(x) (x).GetOrThrow()
+#define ASSERT_PTR(x) (x).GetPtrOrThrow()
+
+#define NB_UNARY_FUNC(func_name, bind_func)                  \
+  static PyObject* func_name(PyObject* self) {               \
+    HANDLE_ERRORS                                            \
+    PyObject* tuple = PyTuple_Pack(1, self);                 \
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor_functions.cpp:32]: 这个tuple需要用PyObjectPtr包一下，因为退出作用域时需要对tuple减引用。下面其他地方都一样需要改一下。

File: oneflow/api/python/framework/tensor.cpp [hunk 3]
```diff
@@ -360,6 +361,21 @@ static PyObject* PyTensorObject__register_storage_delete_hook(PyObject* self, Py
   END_HANDLE_ERRORS
 }
 
+
+PyMethodDef* concat_method_def(PyMethodDef methods[], PyMethodDef extra_methods[]) {
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor.cpp:365]:
```suggestion
static PyMethodDef* concat_method_def(PyMethodDef methods[], PyMethodDef extra_methods[]) {
```

File: oneflow/api/python/framework/tensor.cpp [hunk 4]
```diff
@@ -360,6 +361,21 @@ static PyObject* PyTensorObject__register_storage_delete_hook(PyObject* self, Py
...
+  int len1 = 0;
+  int len2 = 0;
+  PyMethodDef* p1 = methods;
+  PyMethodDef* p2 = extra_methods;
+  while ((p1++)->ml_name != NULL) { len1++; }
+  while ((p2++)->ml_name != NULL) { len2++; }
+  PyMethodDef* total_methods = new PyMethodDef[len1 + len2 + 1];
```
Comment (hjchen2) [inline oneflow/api/python/framework/tensor.cpp:372]: 这里直接new一个PyMethodDef对象返回不大好，因为这个对象并不会被释放。这里最好返回`std::vector< PyMethodDef>`，
```c++
   std::vector< PyMethodDef> total_methods(len1 + len2 + 1);
   for (int i = 0; i < len1; i++) total_methods[i] = methods[i];
   for (int i = 0; i < len2; i++) total_methods[i + len1] = extra_methods[i];
   total_methods[len1 + len2] = {NULL};
   return total_methods;
```
之后在MakeTensorType里面调用concat，
```c++
static PyTypeObject* MakeTensorType() {
  static std::vector<PyMethodDef> PyTensorObject_total_methods = concat_method_def(PyTensorObject_methods, PyTensorObject_extra_methods);
  type->tp_methods = PyTensorObject_total_methods.data();
  ...
)
```

**Review by hjchen2 — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

**Pull request merged**